// @chainlink/belt/lib/src/services/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getJsonFile = exports.debug = exports.getArtifactDirs = exports.getContractDirs = void 0;
const tslib_1 = require("tslib");
const debug_1 = tslib_1.__importDefault(require("debug"));
const fs_1 = require("fs");
const shelljs_1 = require("shelljs");
/**
 * Get contract versions and their directories
 */
function getContractDirs(conf) {
    const contractDirs = shelljs_1.ls(conf.contractsDir);
    return contractDirs
        .map((d) => ({
        dir: d,
        version: conf.compilerSettings.versions[d],
    }))
        .filter((p) => !!p.version);
}
exports.getContractDirs = getContractDirs;
/**
 * Get artifact verions and their directories
 */
function getArtifactDirs(conf) {
    const artifactDirs = shelljs_1.ls(conf.artifactsDir);
    return artifactDirs
        .map((d) => ({
        dir: d,
        version: conf.compilerSettings.versions[d],
    }))
        .filter((p) => !!p.version);
}
exports.getArtifactDirs = getArtifactDirs;
/**
 * Create a logger specifically for debugging. The root level namespace is based on the package name.
 *
 * @see https://www.npmjs.com/package/debug
 * @param fileName The filename that the debug logger is being used in for namespacing purposes.
 */
function debug(fileName) {
    return debug_1.default('belt').extend(fileName);
}
exports.debug = debug;
/**
 * Load a json file at the specified path.
 *
 * @param path The file path relative to the cwd to read in the json file from.
 */
function getJsonFile(path) {
    return JSON.parse(fs_1.readFileSync(path, 'utf8'));
}
exports.getJsonFile = getJsonFile;
