// @chainlink/belt/lib/src/services/compilers/solc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getCompilerOptions = exports.compileAll = void 0;
const sol_compiler_1 = require("@0x/sol-compiler");
const path_1 = require("path");
const utils_1 = require("../utils");
const d = utils_1.debug('solc');
/**
 * Generate solidity artifacts for all of the solidity versions under a specified contract
 * directory.
 *
 * @param conf The application configuration, e.g. where to read solidity files, where to output, etc..
 */
async function compileAll(conf) {
    return Promise.all(utils_1.getContractDirs(conf).map(async ({ dir, version }) => {
        const opts = getCompilerOptions(conf, dir, version);
        const c = new sol_compiler_1.Compiler(opts);
        // Compiler#getCompilerOutputsAsync throws on compilation errors
        // this method prints any errors and warnings for us
        await c.compileAsync();
    }));
}
exports.compileAll = compileAll;
/**
 * Create options for sol-compiler instance that reads in a subdirectory of smart contracts e.g. (src/v0.4, src/v0.5, ..)
 * and outputs their respective compiler artifacts e.g. (abi/v0.4, abi/v0.5)
 *
 * @param config The application specific configuration to use for sol-compiler
 * @param subDir The subdirectory to use as a namespace when reading .sol files and outputting
 * their respective artifacts
 * @param solcVersion The solidity compiler version to use with sol-compiler
 */
function getCompilerOptions({ artifactsDir, useDockerisedSolc, contractsDir, compilerSettings, }, subDir, solcVersion) {
    const _d = d.extend('compiler');
    // remove our custom versions property
    const compilerSettingCopy = JSON.parse(JSON.stringify(compilerSettings));
    // @ts-expect-error
    delete compilerSettingCopy.versions;
    const options = {
        artifactsDir: path_1.join(artifactsDir, subDir),
        compilerSettings: {
            outputSelection: {
                '*': {
                    '*': [
                        'abi',
                        'devdoc',
                        'userdoc',
                        'evm.bytecode.object',
                        'evm.bytecode.sourceMap',
                        'evm.deployedBytecode.object',
                        'evm.deployedBytecode.sourceMap',
                        'evm.methodIdentifiers',
                        'metadata',
                    ],
                },
            },
            ...compilerSettingCopy,
        },
        contracts: '*',
        contractsDir: path_1.join(contractsDir, subDir),
        isOfflineMode: false,
        shouldSaveStandardInput: false,
        solcVersion,
        useDockerisedSolc,
    };
    _d('Compiler options: %o', options);
    return options;
}
exports.getCompilerOptions = getCompilerOptions;
