// @chainlink/belt/lib/src/commands/compile.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const command_1 = require("@oclif/command");
class Compile extends command_1.Command {
    async run() {
        const { flags, args, argv } = this.parse(Compile);
        if (argv.length === 0) {
            this._help();
        }
        try {
            const config = await Promise.resolve().then(() => __importStar(require('../services/config')));
            const conf = config.load(flags.config);
            switch (args.compiler) {
                case 'all':
                    return await this.compileAll('solc', conf);
                case 'all-ovm':
                    return await this.compileAll('solc-ovm', conf);
                default:
                    return await this.compile(args.compiler, conf);
            }
        }
        catch (e) {
            this.error(e);
        }
    }
    async getCompiler(name) {
        return await Promise.resolve().then(() => __importStar(require(`../services/compilers/${name}`)));
    }
    async compile(compilerName, conf) {
        const compiler = await this.getCompiler(compilerName);
        await compiler.compileAll(conf);
    }
    async compileAll(compilerName, conf) {
        const compilers = await Promise.resolve().then(() => __importStar(require('../services/compilers')));
        await this.compile(compilerName, conf);
        await Promise.all([
            compilers.truffle.compileAll(conf),
            compilers.ethers.compileAll(conf),
        ]);
    }
}
exports.default = Compile;
Compile.description = 'Run various compilers and/or codegenners that target solidity smart contracts.';
Compile.examples = [
    `$ belt compile all

Creating directory at abi/v0.4...
Creating directory at abi/v0.5...
Creating directory at abi/v0.6...
Compiling 35 contracts...
...
...
Aggregator artifact saved!
AggregatorProxy artifact saved!
Chainlink artifact saved!
...`,
];
Compile.flags = {
    help: command_1.flags.help({ char: 'h' }),
    config: command_1.flags.string({
        char: 'c',
        default: 'app.config.json',
        description: 'Location of the configuration file',
    }),
};
Compile.args = [
    {
        name: 'compiler',
        description: 'Compile solidity smart contracts and output their artifacts',
        options: ['solc', 'solc-ovm', 'ethers', 'truffle', 'all', 'all-ovm'],
    },
];
